runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/PipelineNode.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.runners.core.construction.graph;
+
+import com.google.auto.value.AutoValue;
+import org.apache.beam.model.pipeline.v1.RunnerApi.PCollection;
+import org.apache.beam.model.pipeline.v1.RunnerApi.PTransform;
+
+/**
+ * A graph node which contains some pipeline element.
+ */
+interface PipelineNode {
```

Comment:
Early on you'll want to figure out that you'll want:
1) Everything that is a node/edge to be cloneable
2) Everything to only implement reference equality

You should make everything only true via reference equality so that during graph manipulation you can clone nodes and insert them back into the graph without it being deduplicated. You can always have libraries which take matchers or equivalence wrappers to find things so losing "actual" equality isn't that bad.

It's best to place these methods so that every implementation inherits them and be very strict when validating this in tests.

Reply:
My claim for (1) is that we'll want to perform some replacement, but I don't know if that's via cloning or via applying a `Node -> Collection<Node>` transformation. Combined with (2), I'm very leery of having two nodes in the graph that are indistinguishable on inspection without access to the graph but report that they are independent of each other (and can coexist). I think we'll need to have some way to distinguish those nodes that will be fairly easy to embed within that node's identity, so we can actually observe both of the nodes and distinguish them ourselves.

With more specific comments on (2), I'd prefer to keep the `Components <-> Network` relationship as intact as possible (so it should be very straightforward to convert back and forth), which means that I think we need to be able to convert a transform or collection into it's associated component without having to traverse the graph to find all nodes that are equal-ish to it.